File: 0127 - Word Ladder/solution.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        from collections import deque
        from string import ascii_lowercase
        res, d, q = 1, set(wordList), deque([beginWord])
        while q:
            n = len(q)
            for _ in range(n):
                s = q.popleft()
                if s == endWord:
                    return res
                for i in range(len(s)):
                    for j in ascii_lowercase:
                        now = s[:i]+j+s[i+1:]
                        if now in d:
                            d.remove(now)
                            q.append(now)
            res += 1
        return 0

File: 0127 - Word Ladder/solution.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> us(wordList.begin(), wordList.end()), h, t;
        if(!us.count(endWord)) return 0;
        h.insert(beginWord);
        t.insert(endWord);
        for(int res = 2; !h.empty()&&!t.empty(); res++){
            if(h.size() > t.size()) swap(h, t);
            unordered_set<string> tmp;
            for(auto s : h){
                for(int i = 0; i < s.size(); i++){
                    char c = s[i];
                    for(char j = 'a'; j <= 'z'; j++){
                        s[i] = j;
                        if(t.count(s)) return res;
                        auto it = us.find(s);
                        if(it != us.end()){
                            tmp.insert(s);
                            us.erase(it);
                        }
                    }
                    s[i] = c;
                }
            }
            swap(h, tmp);
        }
        return 0;
    }
};